src/Lecture3.hs

```diff
@@ -159,10 +182,19 @@ monsters, you should get a combined treasure and not just the first
 🕯 HINT: You may need to add additional constraints to this instance
   declaration.
 -}
-instance Semigroup (Treasure a) where
+instance (Semigroup a) => Semigroup (Treasure a) where
+    (<>) :: Treasure a -> Treasure a -> Treasure a
+    t1 <> t2 =
+        case (t1, t2) of
+          (NoTreasure, SomeTreasure t)     -> SomeTreasure t
+          (SomeTreasure t, NoTreasure)     -> SomeTreasure t
+          (SomeTreasure t, SomeTreasure u) -> SomeTreasure (t <> u)
+          _                                -> NoTreasure
```

Comment:
This is an absolutely correct solution!

However, we can notice that these 4 cases could be reduced to 3 with simpler pattern matching!
Let's see what we can unite in here.

If some of the arguments are `NoTreasure`, you actually always return the second one! So you need to inspect only two cases: `NoTreasure` in the first position, and `NoTreasure` in the second one 🙂

Reply:
I have changed the implementation to this:

```  instance (Eq a, Semigroup a) => Semigroup (Treasure a) where
      (<>) :: Treasure a -> Treasure a -> Treasure a
      t1 <> t2
        | t1 == NoTreasure = t2
        | t2 == NoTreasure = t1
        | otherwise        = SomeTreasure (t <> u)
            where SomeTreasure t = t1
                  SomeTreasure u = t2
```

It works, but the compiler now complains that:
```   instance (Eq a, Semigroup a) => Semigroup (Treasure a) where
      (<>) :: Treasure a -> Treasure a -> Treasure a
      t1 <> t2
        | t1 == NoTreasure = t2
        | t2 == NoTreasure = t1
        | otherwise        = SomeTreasure (t <> u)
            where SomeTreasure t = t1
                  SomeTreasure u = t2
```

Anything wrong that I'm doing?

Reply:
You see this warning because you match on the `t1` value non-exhaustively. The `where` clause is attached to entire guard, not only the latest guard. And `t1` can be `NoTreasure` hence the warning.

This code works because Haskell is lazy and it _doesn't actually perform pattern-matching_ until the `otherwise` case.

A better way of implementing the same logic without GHC warnings would be the following:

```haskell
(<>) :: Treasure a -> Treasure a -> Treasure a
NoTreasure <> x = x
x <> NoTreasure = x
SomeTreasure t <> SomeTreasure u = SomeTreasure (t <> u)
```

Reply:
Just amazing.
I was so focused on guards and `case` and `let ... in` that I forgot about doing without. "Once you have a hammer...". Thanks Dmitrii!